### PR TITLE
when checking whether user has kept a page, also check whether any friends have

### PR DESCRIPTION
**Pull request by 2is10**
@ymatsuda, please take a look. thanks.

### CODE REVIEW

File: shoebox/app/com/keepit/controllers/BookmarksController.scala [hunk 1]
```diff
@@ -123,23 +124,37 @@ object BookmarksController extends FortyTwoController {
...
 
   def checkIfExists(uri: String) = AuthenticatedJsonAction { request =>
-    val bookmark = inject[DBConnection].readOnly { implicit s => 
-      inject[NormalizedURIRepo].getByNormalizedUrl(uri).flatMap { uri =>
-        inject[BookmarkRepo].getByUriAndUser(uri.id.get, request.userId).filter(_.isActive)
+    val userId = request.userId
+    val (uriIdOpt, bookmarkOpt, friendIds) = inject[DBConnection].readOnly { implicit s =>
+      val uriIdOpt = inject[NormalizedURIRepo].getByNormalizedUrl(uri).flatMap(_.id)
+      val bookmarkOpt = uriIdOpt.flatMap { uriId =>
+        inject[BookmarkRepo].getByUriAndUser(uriId, userId).filter(_.isActive)
```
Comment (eishay) [inline shoebox/app/com/keepit/controllers/BookmarksController.scala]: we already do the isActive in the repo

``` scala
  def getByUriAndUser(uriId: Id[NormalizedURI], userId: Id[User])(implicit session: RSession): Option[Bookmark] =
    (for(b <- table if b.uriId === uriId && b.userId === userId && b.state === BookmarkStates.ACTIVE) yield b).firstOption
```

Comment (2is10) [inline shoebox/app/com/keepit/controllers/BookmarksController.scala]: Yes, I didn't add it here. Happy to remove it.

File: shoebox/app/com/keepit/search/graph/URIGraphSearcher.scala [hunk 2]
```diff
@@ -80,6 +80,20 @@ class URIGraphSearcher(searcher: Searcher) {
     }
   }
 
+  def intersectAny(friends: UserToUserEdgeSet, bookmarkUsers: UriToUserEdgeSet): Boolean = {
+    intersectAny(friends.getDestDocIdSetIterator(searcher), bookmarkUsers.getDestDocIdSetIterator(searcher))
+  }
+
+  def intersectAny(i: DocIdSetIterator, j: DocIdSetIterator): Boolean = {
```
Comment (eishay) [inline shoebox/app/com/keepit/search/graph/URIGraphSearcher.scala:88]: cool

File: shoebox/test/com/keepit/search/graph/URIGraphTest.scala [hunk 3]
```diff
@@ -91,8 +91,29 @@ class URIGraphTest extends SpecificationWithJUnit with DbRepos {
   }
   implicit def toSearchable(uriGraphSearcher: URIGraphSearcher) = new Searchable(uriGraphSearcher: URIGraphSearcher)
 
+  class TestDocIdSetIterator(ids: Array[Int]) extends DocIdSetIterator {
```
Comment (gmethvin) [inline shoebox/test/com/keepit/search/graph/URIGraphTest.scala]: You could also do varargs `class TestDocIdSetIterator(ids: Int*) extends DocIdSetIterator` which would make the syntax a bit cleaner. nbd though.

Comment (2is10) [inline shoebox/test/com/keepit/search/graph/URIGraphTest.scala]: I actually wrote that initially, but wasn't sure how to access the varargs by index. Suppose I could have just defined `val idsArray = ids.toArray'. Thanks.

File: shoebox/app/com/keepit/search/graph/URIGraphSearcher.scala (hunk 2 above)
Comment (ymatsuda) [inline shoebox/app/com/keepit/search/graph/URIGraphSearcher.scala:88]: Thanks! It looks good.

**Pull request merged**